drivers/led_strip/ws2812_rpi_pico_pio.c

```diff
@@ -0,0 +1,194 @@
+ */
+RPI_PICO_PIO_DEFINE_PROGRAM(ws2812, 0, 3,
+		/*  .wrap_target                          */
+	0x6221, /*  0: out    x, 1            side 0 [2]  */
+	0x1123, /*  1: jmp    !x, 3           side 1 [1]  */
+	0x1400, /*  2: jmp    0               side 1 [4]  */
+	0xa442, /*  3: nop                    side 0 [4]  */
+		/*  .wrap                                 */
+);
+
+static void ws2812_program_init(PIO pio, uint32_t sm, uint32_t pin, float freq)
+{
+	float clkdiv = sys_clock_hw_cycles_per_sec() / (CYCLES_PER_BIT * freq);
+	uint32_t offset;
+
+	offset = pio_add_program(pio, &ws2812_program);
+
+	pio_sm_config sm_config = pio_get_default_sm_config();
+
+	sm_config_set_wrap(&sm_config, offset + RPI_PICO_PIO_GET_WRAP_TARGET(ws2812),
+			   offset + RPI_PICO_PIO_GET_WRAP(ws2812));
+	sm_config_set_sideset(&sm_config, 1, false, false);
+	sm_config_set_sideset_pins(&sm_config, pin);
+	sm_config_set_out_shift(&sm_config, false, true, 24);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, clkdiv);
+	pio_sm_set_consecutive_pindirs(pio, sm, pin, 1, true);
+	pio_sm_init(pio, sm, offset, &sm_config);
+	pio_sm_set_enabled(pio, sm, true);
+}
+
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
+
+static int ws2812_rpi_pico_pio_update_rgb(const struct device *dev, struct led_rgb *pixels,
+					  size_t num_pixels)
+{
+	const struct ws2812_rpi_pico_pio_config *config = dev->config;
+	struct ws2812_rpi_pico_pio_data *data = dev->data;
+	PIO pio = pio_rpi_pico_get_pio(config->piodev);
+
+	for (size_t i = 0; i < num_pixels; i++) {
+		uint32_t color = 0;
+
+		for (size_t j = 0; j < config->num_colors; j++) {
+			switch (config->color_mapping[j]) {
+			/* White channel is not supported by LED strip API. */
+			case LED_COLOR_ID_WHITE:
+				color |= 0;
+				break;
+			case LED_COLOR_ID_RED:
+				color |= pixels[i].r << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_GREEN:
+				color |= pixels[i].g << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_BLUE:
+				color |= pixels[i].b << (8 * (2 - j));
+				break;
+			}
+		}
+
+		pio_sm_put_blocking(pio, data->sm, color << 8);
+	}
+
+	ws2812_reset_delay(config->reset_delay);
+
+	return 0;
+}
+
+static int ws2812_rpi_pico_pio_update_channels(const struct device *dev, uint8_t *channels,
+					       size_t num_channels)
+{
+	LOG_ERR("update_channels not implemented");
+	return -ENOTSUP;
+}
+
+static const struct led_strip_driver_api ws2812_rpi_pico_pio_api = {
+	.update_rgb = ws2812_rpi_pico_pio_update_rgb,
+	.update_channels = ws2812_rpi_pico_pio_update_channels,
+};
+
+/*
+ * Retrieve the channel to color mapping (e.g. RGB, BGR, GRB, ...) from the
+ * "color-mapping" DT property.
+ */
+static int ws2812_rpi_pico_pio_init(const struct device *dev)
+{
+	const struct ws2812_rpi_pico_pio_config *config = dev->config;
+	struct ws2812_rpi_pico_pio_data *data = dev->data;
```

Comment:
Shouldn't we check here if the PIO device is ready ?

```
 if (!device_is_ready(config->piodev)) {
        LOG_ERR("%s: PIO device not ready", dev->name);
        return -ENODEV;
}
```

Reply:
It's make a sense.
I applied it. Thank you!